#**Aljabar matriks**


1.  penjelasan tentang matriks
2.  operasi matriks
  *   penjumlahan
  *   perkalian
  *   pengurangan
  *   transpose matriks
3. membuat matrik A dan B dengan ordi 3x3 dan matriks C dengan ordo 3x2
4. tugas: buat program untuk mengalikan matriks A dan B tidak menggunakan '@'


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


#Penjelasan Matrik

Matriks adalah sekumpulan bilangan yang disusun berdasarkan baris dan kolom, serta ditempatkan di dalam tanda kurung berupa kurung biasa “( )” atau kurung siku “[ ]”. Suatu matriks diberi nama dengan huruf kapital, seperti A, B, C,dll dan suatu susunan bilangan real atau bilangan kompleks (atau elemen-elemen) yang disusun dalam baris dan kolom sehingga membentuk jajaran persegi panjang.

Ordo dan Elemen Matriks

Matriks itu punya ukuran. Ukuran matriks disebut ordo. Ordo matriks ini berdasarkan dari banyaknya baris dikali banyaknya kolom pada matriks.Ordo matriks merupakan banyaknya baris dan kolom yang terdapat pada suatu matriks. Bentuk umumnya adalah sebagai berikut mxn.Ordo matriks pada matriks  adalah 3x3, yakni terdiri dari 3 baris dan 3 kolom.




masing-masing bilangan yang terdapat di dalam matriks disebut elemen matriks.  matriks dinotasikan dengan huruf kapital, maka elemen-elemen matriks dinotasikan dengan huruf kecil dan diberi indeks yang menyatakan letak baris dan kolomnya.


In [ ]:
import numpy as np
import numpy as sy
a=np.array([[1,2,3],[3,4,3],[5,5,6]])
print(a)

[[1 2 3]
 [3 4 3]
 [5 5 6]]


$$
\begin{equation*}
A=
\begin{bmatrix}
1&2&3\\
3&4&3\\5&5&6
\end{bmatrix}
\end{equation*}
$$

#OPERASI PADA MATRIKS

In [ ]:
#A dan B= 3X3
#C = 3X2

import numpy as np
import sympy as sy
a=np.array ([[1, 0, 2],[2, 0, 4],[4, 6, 8]])
b=np.array ([[2, 3, 4],[4, 5, 6], [1, 2, 4]])
c=np.array ([[2, 3],[4, 5],[6, 7]])

print('A=\n',a)
print('B=\n',b)
print('C=\n',c)


A=
 [[1 0 2]
 [2 0 4]
 [4 6 8]]
B=
 [[2 3 4]
 [4 5 6]
 [1 2 4]]
C=
 [[2 3]
 [4 5]
 [6 7]]


$$
\begin{align*}
A=
\begin{bmatrix}
1&0&2\\
2&0&4\\
4&6&8
\end{bmatrix}
&&
B=
\begin{bmatrix}
2&3&4\\
4&5&6\\1&2&4
\end{bmatrix}
&&
C=
\begin{bmatrix}
2&3\\
4&5\\
6&7
\end{bmatrix}
\end{align*}
$$


1. Penjumlahan Matrik

Definisi: jika A dan B adalah matriks-matriks berukuran (berdimensi) sama, maka
jumlah A+B adalah matriks yang diperoleh dengan menambahkan elemen-elemen A
dengan elemen-elemen B yang letaknya bersesuaian.
Dua matriks dapat dijumlahkan jika jumlah baris dan kolomnya sama. Matriks hasil penjumlahannya juga akan memiliki ordo yang sama.

In [ ]:
print('A+B=\n',a+b)
print('B+A=\n',b+a)
print('sifat di atas merupakan sifat komunitatif karena A+B = B+A')

A+B=
 [[ 3  3  6]
 [ 6  5 10]
 [ 5  8 12]]
B+A=
 [[ 3  3  6]
 [ 6  5 10]
 [ 5  8 12]]
sifat di atas merupakan sifat komunitatif karena A+B = B+A


$$
\begin{align*}
A+B=
\begin{bmatrix}
3&3&6\\
6&5&10\\
5&8&12
\end{bmatrix}
&&
B+A=
\begin{bmatrix}
3&3&6\\
6&5&10\\
5&8&12
\end{bmatrix}
\end{align*}
$$


sifat di atas merupakan sifat komutatif karena A+B = B+A

2.Pengurangan matrik

Defenisi: selisih A-B adalah matriks
yang diperoleh dengan mengurangkan elemen-elemen A dengan elemen-elemen B
yang letaknya bersesuaian. Pengurangan matrik memiliki konsep yang sama dengan penjumlahan. Dua buah matriks dapat dikurangkan apabila keduanya memiliki ordo yang sama.

In [ ]:
print('A-B=\n',a-b)
print('B-A=\n',b-a)
print('sifat di atas merupakan sifat komutatif karena A-B = B-A')

A-B=
 [[-1 -3 -2]
 [-2 -5 -2]
 [ 3  4  4]]
B-A=
 [[ 1  3  2]
 [ 2  5  2]
 [-3 -4 -4]]
sifat di atas merupakan sifat komutatif karena A-B = B-A


\begin{align*}
A-B=
\begin{bmatrix}
-1&-3&-2\\
-2&-5&-2\\
3&4&4
\end{bmatrix}
&&
B-A=
\begin{bmatrix}
1&3&2\\
2&5&2\\
-3&-4&-4
\end{bmatrix}
\end{align*}

sifat di atas merupakan sifat komutatif karena A-B = B-A


3.Perkalian Matriks

Definisi: Perkalian matriks sendiri adalah proses mengalikan setiap elemen baris pada matriks pertama dengan elemen kolom pada matriks kedua. Hasilnya kemudian dijumlahkan untuk memperoleh hasil akhir perkalian matriks atau Kalikan elemen-elemen yang bersesuaian dari
baris dan kolom secara bersama-sama dan kemudian jumlahkan hasil kalinya.

In [ ]:
print('AB=\n',a@b)
print('AC=\n',a@c)
print('BC=\n',b@c)

AB=
 [[ 4  7 12]
 [ 8 14 24]
 [40 58 84]]
AC=
 [[14 17]
 [28 34]
 [80 98]]
BC=
 [[40 49]
 [64 79]
 [34 41]]


$$
\begin{align*}
AB=
\begin{bmatrix}
4&7&12\\
8&14&24\\
40&58&84
\end{bmatrix}
&&
AC=
\begin{bmatrix}
14&17\\
28&34\\
80&98
\end{bmatrix}
&&
BC=
\begin{bmatrix}
40&49\\
64&79\\
34&41
\end{bmatrix}
\end{align*}
$$


*  Sifat Asosiasi:

sifat asosiatif adalah kondisi ketika operasi hitung tiga angka atau lebih, hasilnya tidak bergantung pada pengelompokan dari angka yang dioperasikan.

 (AB)C = A(BC)


In [ ]:
print('A(BC) =\n', a@(b@c))
print('(AB)C =\n',(a@b)@c)

A(BC) =
 [[108 131]
 [216 262]
 [816 998]]
(AB)C =
 [[108 131]
 [216 262]
 [816 998]]


\begin{equation*}
A(BC)=
\begin{bmatrix}
108&131\\
216&262\\816&998
\end{bmatrix}
(AB)C=
\begin{bmatrix}
108&131\\
216&262\\
816&998
\end{bmatrix}
\end{equation*}

*  perkalian skalar

 proses mengalikan suatu matriks dengan skalar. Misalkan 'A' adalah matriks dan 'c' adalah skalar (bilangan real).

  c(AB) = (cA)B = A(cB)

In [ ]:
c = 3
print('c(AB)\n',c*(a@b))
print('(cA)B\n',(c*a)@b)
print('A(cB)\n',a@(c*b))

c(AB)
 [[ 12  21  36]
 [ 24  42  72]
 [120 174 252]]
(cA)B
 [[ 12  21  36]
 [ 24  42  72]
 [120 174 252]]
A(cB)
 [[ 12  21  36]
 [ 24  42  72]
 [120 174 252]]


\begin{equation*}
c(AB)=
\begin{bmatrix}
12&21&36\\
24&42&72\\120&174&252
\end{bmatrix}
(cA)B=
\begin{bmatrix}
12&21&36\\
24&42&72\\
120&174&252
\end{bmatrix}
A(cB)=
\begin{bmatrix}
12&21&36\\
24&42&72\\
120&174&252
\end{bmatrix}
\end{equation*}

*  perkalian identitas

matriks persegi yang setiap elemen diagonal utamanya bernilai 1 dan elemen lainnya bernilai 0.

  AI = IA = A

In [ ]:
i = np.array([[1,0,0],[0,1,0],[0,0,1]])
print('AI=\n',a@i)
print('IA=\n',i@a)
print('A=\n',a)

AI=
 [[1 0 2]
 [2 0 4]
 [4 6 8]]
IA=
 [[1 0 2]
 [2 0 4]
 [4 6 8]]
A=
 [[1 0 2]
 [2 0 4]
 [4 6 8]]


$$
\begin{equation*}
AI=
\begin{bmatrix}
1&0&2\\
2&0&4\\4&6&8
\end{bmatrix}
IA=
\begin{bmatrix}
1&0&2\\
2&0&4\\
4&6&8
\end{bmatrix}
A=
\begin{bmatrix}
1&0&2\\
2&0&4\\
4&6&8
\end{bmatrix}
\end{equation*}
$$

*  tidak berlaku sifat kumutatif

Sifat komutatif perkalian menyatakan bahwa meskipun urutan bilangan diubah, hasil kali dua bilangan bulat atau lebih tetap sama.

   AB != BA atau AB tidak sama dengan BA

In [ ]:
print('AB=\n',a@b)
print('BA=\n',b@a)

AB=
 [[ 4  7 12]
 [ 8 14 24]
 [40 58 84]]
BA=
 [[24 24 48]
 [38 36 76]
 [21 24 42]]


\begin{align*}
AB=
\begin{bmatrix}
4&-7&4\\
8&14&24\\
40&58&84
\end{bmatrix}
&&
BA=
\begin{bmatrix}
24&24&48\\
38&36&76\\
21&24&76
\end{bmatrix}
\end{align*}



##4. transpose matriks


Transpose matriks adalah operasi yang menghasilkan matriks baru dengan menukar baris dan kolom dari matriks asalnya. Dengan kata lain, jika matriks asalnya memiliki dimensi \(mxn), maka transpose-nya akan memiliki dimensi \(nxm), di mana elemen-elemennya dipertukarkan antara baris dan kolom sesuai dengan posisi aslinya.

In [ ]:
c=np.array([[2, 3],[4, 5],[6, 7]])
c_tran = c.transpose()
print('matrik c sebelum di transpose=\n', c)
print('matrik c setelah di transpose=\n',c_tran)
print('ordo C_transpose adalah',c_tran.shape)
z=c_tran@c
print('Z = C_transpose x C =\n',z)
print('ordo z adalah',z.shape)
p=c@c_tran
print('P = C_transpose x C =\n',p)
print('ordo P adalah',p.shape)
print("Dapat disimpulkan bahwa C_transpose X C tidak sama dengan C X C_transpose")

matrik c sebelum di transpose=
 [[2 3]
 [4 5]
 [6 7]]
matrik c setelah di transpose=
 [[2 4 6]
 [3 5 7]]
ordo C_transpose adalah (2, 3)
Z = C_transpose x C =
 [[56 68]
 [68 83]]
ordo z adalah (2, 2)
P = C_transpose x C =
 [[13 23 33]
 [23 41 59]
 [33 59 85]]
ordo P adalah (3, 3)
Dapat disimpulkan bahwa C_transpose X C tidak sama dengan C X C_transpose


##program untuk mengalikan matriks

In [ ]:
def input_matrix():
    baris = int(input("Masukkan jumlah baris: "))
    kolom = int(input("Masukkan jumlah kolom: "))
    matrix = []
    for i in range(baris):
        row = []
        print(f"Masukkan elemen untuk baris {i + 1}: ")
        for j in range(kolom):
            element = float(input(f"Elemen [{i + 1}, {j + 1}]: "))
            row.append(element)
        matrix.append(row)
    return matrix

def perkalian(matrix1, matrix2):
    if len(matrix1[0]) != len(matrix2):
        print("Perkalian matriks tidak dapat dilakukan karena jumlah kolom dalam matriks pertama tidak sama dengan jumlah baris dalam matriks kedua")
        return None
    result = [[0 for _ in range(len(matrix2[0]))] for _ in range(len(matrix1))]
    for i in range(len(matrix1)):
        for j in range(len(matrix2[0])):
            for k in range(len(matrix2)):
                result[i][j] += matrix1[i][k] * matrix2[k][j]
    return result

def tampilan_hasil(matrix):
    for baris in matrix:
        print(baris)
coba = input("Ingin mengisi matriks sendiri? (iya/tidak) : ")
if coba == "iya":
    print("Program perkalian matriks!!")
    print("Masukkan matriks pertama:")
    matriks1 = input_matrix()
    print ("Matriks 1 = ", matriks1)
    print("Masukkan matriks kedua:")
    matriks2 = input_matrix()
    print ("Matriks 2 = ",matriks2)
elif coba == "tidak":
    print ("Berikut adalah contoh matriks")
    print ("""
    matriks 1=([[1, 0, 2],[2, 0, 4],[4, 6, 8]])
    matriks 224=([[2, 3, 4],[4, 5, 6], [1, 2, 4]])
    """)
    matriks1=([[1, 0, 2],[2, 0, 4],[4, 6, 8]])
    matriks2=([[2, 3, 4],[4, 5, 6], [1, 2, 4]])
else:
    print("Maaf pilihan ersebut tidak tersedia. Tolong masukkan pilihan yang sesuai")

test = perkalian(matriks1, matriks2)
print("\nHasil perkalian matriks:")
tampilan_hasil(test)


Ingin mengisi matriks sendiri? (iya/tidak) : iya
Program perkalian matriks!!
Masukkan matriks pertama:
Masukkan jumlah baris: 2
Masukkan jumlah kolom: 2
Masukkan elemen untuk baris 1: 
Elemen [1, 1]: 3
Elemen [1, 2]: 2
Masukkan elemen untuk baris 2: 
Elemen [2, 1]: 4
Elemen [2, 2]: 2
Matriks 1 =  [[3.0, 2.0], [4.0, 2.0]]
Masukkan matriks kedua:
Masukkan jumlah baris: 2
Masukkan jumlah kolom: 2
Masukkan elemen untuk baris 1: 
Elemen [1, 1]: 3
Elemen [1, 2]: 1
Masukkan elemen untuk baris 2: 
Elemen [2, 1]: 4
Elemen [2, 2]: 2
Matriks 2 =  [[3.0, 1.0], [4.0, 2.0]]

Hasil perkalian matriks:
[17.0, 7.0]
[20.0, 8.0]
